patch(DPE-7576): New MongoDB 6.0.24 snap

Gu1nness opened the pull request:
<!-- Provide a general summary of your changes in the Title above -->
<!-- markdownlint-disable MD041 -->

## 🏷 Type of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation Update
- [ ] Tooling and CI changes
- [x] Dependencies upgrade or change

## 📝 Description
<!--
Describe your changes in detail including:
  - the purpose and scope of this PR (what)
  - the impacted components (where, often match conventional commit scope)
  - explanation/algorithm if require (how)
-->
This PR bumps the version of the snap and rock to use the latest MongoDB 6 revision: 6.0.24

## 📑 Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- Either the associated JIRA reference or a paragraph. -->
Prepare the ground for a new stable release and for the major upgrade from Mongodb 6 to 8
CF [this JIRA ticket](https://docs.google.com/document/d/1GU_05gjTC4MvVERqm8Oai7buGLUtrw4ooxMQXV0BwGE/edit?usp=sharing)

## 🧪 How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Integration tests here, snap tested and rock tested before.

## ✅ Checklist
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING**](../blob/6/edge/CONTRIBUTING.md) document.
- [ ] I have added tests to cover my changes.
- [x] All new and existing tests passed.

Gu1nness committed:
feat(DPE-7576): New MongoDB 6.0.24 snap

Gu1nness committed:
fix(backups): Update config workflow for PBM 2.9

With PBM 2.9, a few changes are required to configure PBM.
First, it's not possible anymore to use an empty yaml file to setup the
configuration.
Moreover, it is still necessary to initialise the configuration from the
configuration file during the bootstrap.
Hence this solution:
 * Boostrap with a blackhole storage (which is a noop storage for backups).
 * To avoid resync being triggered multiple times during the
   configuration setting, we now set all keys at once. resync is
   triggered automatically everytime a key from a storage config is set
   so it tends to overload the service and break the next keys from
   being set.

Gu1nness committed:
fix extend timeout even more

Gu1nness committed:
fix: replica_info has changed

Gu1nness committed:
fix: replica_info has changed

Gu1nness committed:
fix: release after CI

Gu1nness committed:
feat: S3 Bucket creation logic

Gu1nness committed:
fix: InvalidAccessKeyId

Gu1nness committed:
fix: endpoint is mandatory

Gu1nness committed:
fix: SignatureDoesNotMatch as well

Gu1nness committed:
fix: set credentials and precedence of configuration over bucket creation

Gu1nness committed:
fix(backups): Location constraint

If a region is provided, we must exclude the case where the region is
`us-east-1`. This default value is not allowed and will raise an error
when trying to create the bucket, EVEN when the bucket already exists.

cf: https://docs.aws.amazon.com/AmazonS3/latest/API/API_CreateBucket.html#API_CreateBucket_RequestBody
cf: https://github.com/aws/aws-sdk-js/issues/3647
cf: https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/S3.html#createBucket-property

Gu1nness committed:
fix: Don't try to create if not allowed to do so

Gu1nness committed:
fix: Reset statuses after resync

Gu1nness committed:
fix: Wait before set config in case a resync is in progress

Gu1nness committed:
fix: log line

Gu1nness committed:
Revert "fix: Wait before set config in case a resync is in progress"

This reverts commit 2b95e9d9639e03c1ff2eb9767a7093bac3eaae18.

Gu1nness committed:
fix: Don't restart on resync

Gu1nness committed:
fix: wait for backup to succeed before restoring

Mehdi-Bendriss commented on single_kernel_mongo/config/statuses.py:
```diff
@@ -116,13 +116,29 @@ class BackupStatuses(Enum):
     PBM_INCOMPATIBLE_CONF = StatusObject(
-        status="blocked", message="s3 configurations are incompatible."
+        status="blocked",
+        message="s3 configurations are incompatible.",
```
```suggestion
        message="s3 config options are incompatible.",
```

Mehdi-Bendriss commented on single_kernel_mongo/config/statuses.py:
```diff
@@ -116,13 +116,29 @@ class BackupStatuses(Enum):
+    FAILED_TO_CREATE_BUCKET = StatusObject(
+        status="blocked",
+        message="Failed to create buckets, check logs.",
+        action="Check S3 configuration",
```
```suggestion
        action="Check S3 configuration on s3-integrator.",
```

Mehdi-Bendriss commented on single_kernel_mongo/events/backups.py:
```diff
@@ -119,12 +121,41 @@ def _on_s3_credential_changed(self, event: CredentialsChangedEvent) -> None:  #
             return
 
+        if not self.manager.validate_s3_config():
+            logger.info(
```
Probably `.warning` ?

Gu1nness committed:
fix: waiting status if resync is in progress

Mehdi-Bendriss commented on single_kernel_mongo/managers/backups.py:
```diff
@@ -141,6 +153,92 @@ def backup_id(self) -> str:
+                bucket.create()
+            bucket.wait_until_exists()
+        except ClientError as e:
+            if (
+                # AWS returns these if the bucket was already created
+                "BucketAlreadyOwnedByYou" in e.args[0]
+                or "BucketAlreadyExists" in e.args[0]
+                # GCP returns this if the bucket was already created
+                or "BucketNameUnavailable" in e.args[0]
```
when would this ever happen? given you're already checking on the existence of the bucket earlier and return if it exists

Gu1nness replied:
I forgot to remove this one, in the first code I wrote it was not checking for the bucket existence before trying to create it.
Let me remove those checks.

Mehdi-Bendriss commented on single_kernel_mongo/events/backups.py:
```diff
@@ -119,14 +121,42 @@ def _on_s3_credential_changed(self, event: CredentialsChangedEvent) -> None:  #
             self.manager.set_config_options(credentials=credentials)
+            # Finally, resync the configuration.
             self.manager.resync_config_options()
+        except InvalidS3CredentialsError:
```
We are missing the handling of a couple of exceptions here `SSLError, ConnectionTimeoutError`

Gu1nness replied:
So more statuses ? Or we let the charm error, let the user fix and then resolve ?

Mehdi-Bendriss replied:
I think we should log those and treat them as `FAILED_TO_CREATE_BUCKET` exceptions

Gu1nness replied:
Ok  !

Mehdi-Bendriss commented on single_kernel_mongo/managers/backups.py:
```diff
@@ -141,6 +153,92 @@ def backup_id(self) -> str:
+            bucket.meta.client.head_bucket(Bucket=bucket_name)
+            logger.info(f"Using existing bucket {bucket_name}")
+            exists = True
+        except ConnectTimeoutError as e:
```
This is worth putting in a Tenacity retry context manager with a few second wait, to potentially resolve the `ConnectionTimeoutError`

Mehdi-Bendriss commented on single_kernel_mongo/managers/backups.py:
```diff
@@ -308,6 +412,14 @@ def backup_state(self) -> BackupState:
+            self.create_bucket(credentials)
+        except InvalidS3CredentialsError:
+            return BackupState.INCORRECT_CREDS
+        except FailedToCreateS3BucketError:
```
We are missing the handling of a couple of exceptions here `SSLError, ConnectionTimeoutError`

Mehdi-Bendriss commented on single_kernel_mongo/managers/backups.py:
```diff
@@ -426,7 +540,8 @@ def validate_s3_config(self) -> bool:
         if not provided_configs.get("storage.s3.endpointUrl"):
-            logger.info("Missing endpointUrl - this is required for MinIO and GCP")
+            logger.info("Missing s3 endpoint.")
+            return False
```
If I understand correctly, this should be False only in the case of MinIO and GCP, not aws - in which case the config will never be valid even it technically is.

Gu1nness replied:
We rely on the endpoint in the bucket creation even for AWS so I'd rather make it mandatory (and have the same level of checks as etcd charm or postgres charm (for example) than having an exception for AWS. I don't now why it was made optional for AWS while PBM does not make it optional for AWS.

Mehdi-Bendriss commented on single_kernel_mongo/managers/backups.py:
```diff
@@ -676,9 +809,13 @@ def _wait_pbm_status(self) -> None:
         except WorkloadExecError as e:
             if status := self.process_pbm_error_as_status(e.stdout):
                 self.state.statuses.add(status, scope="unit", component=self.name)
+                return
```
Should we not raise if no status?

Mehdi-Bendriss commented on tests/integration/mongodb/backups/test_backups.py:
```diff
@@ -102,31 +107,15 @@ async def test_blocked_incorrect_creds(ops_test: OpsTest, substrate: Substrate)
 
 
-@pytest.mark.abort_on_fail
-async def test_blocked_incorrect_conf(
```
why is this test no longer relevant?

Gu1nness replied:
This test was designed to check if we succesfully ended up in the "s3 configurations are incompatible", which was raised if the bucket was not existing (because we were under the assumption that the operator had to provide a working bucket).
Since we now create the bucket before anything and will not proceed further if the bucket is not created, it's not needed anymore.

Mehdi-Bendriss commented on single_kernel_mongo/managers/backups.py:
```diff
@@ -141,6 +153,92 @@ def backup_id(self) -> str:
+                request_checksum_calculation="when_required",
+                response_checksum_validation="when_required",
+            ),
+            verify=(TRUST_STORE_PATH / TrustStoreFiles.PBM.value)
```
Could we replace the value of the `TRUST_STORE_PATH` with a localized path? (withing `SNAP_DATA`)

Gu1nness replied:
On VM that path is already used to store the SSL certificate, this is specifically important for k8s where we have to land the certificate on the charm container.
I can move it to another location but I'm not sure where and if that would make sense or not?

Mehdi-Bendriss replied:
for k8s it should be fine, however the path on VM is still `/usr/local/share/ca-certificates` and not relative to `SNAP_DATA`

Gu1nness replied:
Yes but we have not decided on an implementation for those paths and this addition would imply ours of development and new snap + rock releases to add the correct environment variables.
When I say "On VM that path is already used" it means that this is the path that we're currently using when storing that exact certificate for PBM anyway.
I had asked if there would be a need for such a spec/implementation for certificates in a Data Office Hour and the answer seemed to be that there was no need for that from mykola.

Mehdi-Bendriss replied:
The issue with `/usr/local/share/ca-certificates` is that the new certificate or CA is trusted system wide, not only by mongodb. This should be our last recourse only IF pbm doesn't accept custom trust stores.

I am fine moving forward as-is for the time being, but we need to investigate more profoundly how to make PBM trust certificates located in a non-system wide trust store.

Let's track it [here](https://github.com/canonical/mongo-single-kernel-library/issues/54)

Gu1nness replied:
PBM use the system certificates.
So to use a "local trust store" for the snap would mean:
 * Update snap to create `/var/snap/charmed-mongodb/current/etc/ssl/certs` + prepopulate that directory.
 * In the snap, add an environment variable to PBM and PBM agent that specify to look for certificates in `/var/snap/charmed-mongodb/current/etc/ssl/certs`.
 * Update the calls to `/update-ca-certificates` to add `--etccertsdir /var/snap/charmed-mongodb/current/etc/ssl/certs`
 * Optionally also create `/var/snap/charmed-mongodb/current/usr/local/share/ca-certificates` and same add the right argument to the `update-ca-certificates` call

Mehdi-Bendriss replied:
This would be the way, we have introduced a [spec](https://docs.google.com/document/d/1dvy2sEkCliSFLAXKBkGiyMuJCE6wiO794N3HkaDNGfE/edit?tab=t.0) for jvm based databases on the same topic

Mehdi-Bendriss reviewed (commented):
Thanks Neha! Great work! I left a few comments

Gu1nness committed:
fix: first round of reviews

Gu1nness committed:
feat: Handle ConnectTimeout and SSLError

Gu1nness committed:
Apply suggestions from code review

Co-authored-by: Mehdi Bendriss <bendrissmehdi@gmail.com>

Gu1nness committed:
feat: formattingé

Gu1nness committed:
fix: unit test

Mehdi-Bendriss reviewed (dismissed):
Thanks Neha!

Gu1nness committed:
fix: bump DPH to fix action bug

Gu1nness committed:
fix: unit test

Gu1nness committed:
fix: Update libs (specifically for glauth)

Gu1nness committed:
fix: restore ldap lib v10 to match released charm

Gu1nness committed:
fix: inverted check

Gu1nness committed:
fix: revert library revision that adds incompatibility between mongos and mongodb

Gu1nness committed:
fix: Try new revision

Gu1nness committed:
fix: Try to change integration order

Gu1nness committed:
fix: list

Gu1nness committed:
fix: Deploy postgres first + use tmate

Gu1nness committed:
fix: one more round

Gu1nness committed:
fix: on success

Gu1nness committed:
fix: remove context manager

Gu1nness committed:
fix: storage pgdata size

Gu1nness committed:
fix: tmate on failure

Gu1nness committed:
fix: reduce number of units deployed

Gu1nness committed:
fix: disable tmate

Gu1nness committed:
fix: upgrade + cleanup space

Gu1nness committed:
fix: better cleanup

Gu1nness committed:
fix: use cache for side charms

Gu1nness committed:
fix: swift not switft

Gu1nness committed:
fix: Use DPH 0.1.7 + update helpers for json output

Gu1nness committed:
fix: Prevent PrematureDataAccessError

Gu1nness committed:
fix: More time for free up disk space

Gu1nness committed:
feat: Delays cluster relation establishment

adds: checks on databag before writing for cluster relation
bumps: data_interfaces to libpatch 53

Gu1nness committed:
fix: charm version wrong

Mehdi-Bendriss commented on single_kernel_mongo/config/statuses.py:
```diff
@@ -116,13 +116,29 @@ class BackupStatuses(Enum):
     ACTIVE_IDLE = StatusObject(status="active", message="")
+    FAILED_TO_CREATE_BUCKET = StatusObject(
+        status="blocked",
+        message="Failed to create buckets, check logs.",
```
```suggestion
        message="Failed to create S3 bucket, check logs.",
```

Mehdi-Bendriss commented on single_kernel_mongo/events/backups.py:
```diff
@@ -119,14 +122,43 @@ def _on_s3_credential_changed(self, event: CredentialsChangedEvent) -> None:  #
+            event.defer()
+            return
         except SetPBMConfigError:
-            logger.error("Failed to configure s3 backup options")
```
why not keep?

Gu1nness replied:
It's already logged when the error is raised with more information. Prevent duplication of logs.

Gu1nness committed:
fix: safeguard if the relation was never fully established for cluster, nothing to clean

Mehdi-Bendriss commented on single_kernel_mongo/managers/backups.py:
```diff
@@ -141,6 +153,89 @@ def backup_id(self) -> str:
+        try:
+            bucket.meta.client.head_bucket(Bucket=bucket_name)
+            logger.info(f"Using existing bucket {bucket_name}")
+            exists = True
```
this is not needed, we can return here, and omit the need for this variable altogether

Mehdi-Bendriss commented on single_kernel_mongo/managers/backups.py:
```diff
@@ -141,6 +153,89 @@ def backup_id(self) -> str:
+            logger.error(f"error: {e!s} - Is TLS enabled and CA chain set on S3?")
+            raise e
+
+        if exists:
+            return
```
related to the above comment

Mehdi-Bendriss commented on single_kernel_mongo/managers/backups.py:
```diff
@@ -388,7 +499,7 @@ def resync_config_options(self):  # pragma: nocover
                         raise PBMBusyError
                     case BackupState.WAITING_TO_SYNC:
-                        self.workload.restart()
+                        # self.workload.restart()
```
Can we delete?

Mehdi-Bendriss commented on single_kernel_mongo/managers/mongo.py:
```diff
@@ -274,18 +274,22 @@ def add_user(self, relation: Relation):
+
         with MongoConnection(self.state.mongo_config) as mongo:
             has_user = mongo.user_exists(username)
-        if has_user:
```
This was a funny bug, glad you found it

Gu1nness replied:
This has been raised by Mateo !
He realised he was lacking the access to his user but it had the accesses to the default database.
But this is good that it's fixed :)

Mehdi-Bendriss reviewed (commented):
Thank you Neha! Really good job here! 
I only have a few minor comments and we should be good

Gu1nness committed:
fix: last changes after review
